library/lib_window_menu.ks

```diff
@@ -0,0 +1,87 @@
+// This script shows a menu on the terminal allowing user to
+// select one of the options and return it to calling script.
+@lazyglobal off.
+
+run lib_window.
+
+function open_window_menu{
+	parameter title.
+	parameter list_of_names.
+	parameter window.
+
+	local x is window[0].
+	local y is window[1].
+	
+	local current_option is 0.
+	local len is list_of_names:length().
+
+	local last_up is ag7.
+	local last_down is ag8.
+	local last_sel is ag9.
+	local process_state is list(
+		list(false,window,"update_window_menu",false),
+		current_option,last_up,last_down,last_sel,list_of_names,title
```

Comment:
Actually I considered all three ways (the third one being just using a single list for everything). The reason why I chose the my way is because if window dependent state was inner list, we would have another layer of lists everywhere. And as you mentioned above, the code is barely readable already. I'd like to stick with "system process state" list, because we can write generic access functions, such as get_process_window, or is_process_finished, which would abstract it away. Generic code cannot help, by definition, with type-specific state.